### PR TITLE
perf: allow reco for DD4hep detector

### PR DESCRIPTION
**Pull request by XiaocongAi**
Currently, the track reco is using sim hits read in by the `CsvPlanarClusterReader`, but the `PlanarSteppingAlgorithm` is only working for the `GenericDetector`. Hence we only have clusters info for the `GenericDetector`, which blocks the track reco for other detectors.

This PR replaces the `CsvPlanarClusterReader` with `CsvSimHitReader` to read in directly the sim hits (written out by the `CsvSimHitWriter` after simulation) for the track reco. The sim hits will be passed to hit smearing for further processing.

It also implements a common interface for track reco with different detectors and add optionally the reco executables for DD4hep detector.

The pull of fitted perigee track parameters for the Open Data Detector is as below:
image

### CODE REVIEW

**Comment by asalzburger** [comment]
Great to see this appearing.

File: Examples/Run/Reconstruction/Common/RecCKFTracks.cpp [hunk 1]
```diff
@@ -32,9 +32,8 @@
 using namespace Acts::UnitLiterals;
 using namespace ActsExamples;
 
-int main(int argc, char* argv[]) {
-  GenericDetector detector;
-
+int recCKFTracks(int argc, char* argv[],
```
Comment (msmk0) [inline Examples/Run/Reconstruction/Common/RecCKFTracks.cpp:35]:
```suggestion
int runRecCKFTracks(int argc, char* argv[],
```
Suggestion: naming similar to e.g. the `runFatras` main method

File: Examples/Run/Reconstruction/Common/RecTruthTracks.cpp [hunk 2]
```diff
@@ -33,9 +33,8 @@
 using namespace Acts::UnitLiterals;
 using namespace ActsExamples;
 
-int main(int argc, char* argv[]) {
-  GenericDetector detector;
-
+int recTruthTracks(int argc, char* argv[],
```
Comment (msmk0) [inline Examples/Run/Reconstruction/Common/RecTruthTracks.cpp:36]:
```suggestion
int runRecTruthTracks(int argc, char* argv[],
```
Suggestion: as above.

File: Examples/Run/Reconstruction/DD4hep/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,22 @@
+add_executable(
+  ActsExampleTruthTracksDD4hep
+  DD4hepRecTruthTracks.cpp)
+
+target_link_libraries(
+	ActsExampleTruthTracksDD4hep
```
Comment (msmk0) [inline Examples/Run/Reconstruction/DD4hep/CMakeLists.txt:6]:
```suggestion
  ActsExampleTruthTracksDD4hep
```
Whitespace.

File: Examples/Run/Reconstruction/DD4hep/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  PRIVATE
+  ActsExamplesRecTracksCommon ActsExamplesDetectorDD4hep)
+
+add_executable(
+    ActsExampleCKFTracksDD4hep
+    DD4hepRecCKFTracks.cpp)
+target_link_libraries(
+    ActsExampleCKFTracksDD4hep
+    PRIVATE
+  ActsExamplesRecTracksCommon ActsExamplesDetectorDD4hep)
```
Comment (msmk0) [inline Examples/Run/Reconstruction/DD4hep/CMakeLists.txt:16]:
```suggestion
add_executable(
  ActsExampleCKFTracksDD4hep
  DD4hepRecCKFTracks.cpp)
target_link_libraries(
  ActsExampleCKFTracksDD4hep
  PRIVATE
    ActsExamplesRecTracksCommon
    ActsExamplesDetectorDD4hep)
```
Whitespace

File: Examples/Run/Reconstruction/DD4hep/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+install(
+  TARGETS
+  ActsExampleTruthTracksDD4hep
+  ActsExampleCKFTracksDD4hep
```
Comment (msmk0) [inline Examples/Run/Reconstruction/DD4hep/CMakeLists.txt:21]:
```suggestion
    ActsExampleTruthTracksDD4hep
    ActsExampleCKFTracksDD4hep
```
Whitespace

**Review by msmk0 — APPROVED** [review]
@XiaocongAi Nice, straightforward PR. I only have minor suggestions; please have a look if you want to include them or not.

**Comment by XiaocongAi** [comment]
Hi @msmk0 , thanks! Sure, I will address them a bit later.

File: Examples/Run/Reconstruction/Common/RecCKFTracks.cpp (hunk 1 above)
Comment (XiaocongAi) [reply Examples/Run/Reconstruction/Common/RecCKFTracks.cpp]: Thanks. Changed.

File: Examples/Run/Reconstruction/Common/RecTruthTracks.cpp (hunk 2 above)
Comment (XiaocongAi) [reply Examples/Run/Reconstruction/Common/RecTruthTracks.cpp]: Changed.

File: Examples/Run/Reconstruction/DD4hep/CMakeLists.txt (hunk 3 above)
Comment (XiaocongAi) [reply Examples/Run/Reconstruction/DD4hep/CMakeLists.txt]: I don't realize that the format here has been fixed manually.. Fixed.

File: Examples/Run/Reconstruction/DD4hep/CMakeLists.txt (hunk 4 above)
Comment (XiaocongAi) [reply Examples/Run/Reconstruction/DD4hep/CMakeLists.txt]: Fixed.

File: Examples/Run/Reconstruction/DD4hep/CMakeLists.txt (hunk 5 above)
Comment (XiaocongAi) [reply Examples/Run/Reconstruction/DD4hep/CMakeLists.txt]: Fixed.

**Comment by XiaocongAi** [comment]
Hi @msmk0 , I should have made the changes as suggested.

**Review by msmk0 — APPROVED** [review]

**Pull request merged**